File: ejemplos/02-pub-sub-basico/suscriptor.py
import time
import sys
from nerve import NexusClient

# Tomamos el nombre del suscriptor por argumento de consola, 
# para poder ejecutar varios distintos. 
# Si no hay argumento, le ponemos uno por defecto.
nombre_nodo = sys.argv[1] if len(sys.argv) > 1 else "suscriptor_default"

def al_recibir_noticia(mensaje_crudo):
    # En un patrón Pub-Sub usando 'broadcast', todos reciben el mensaje.
    # El trabajo del suscriptor es FILTRAR lo que le interesa.
    payload = mensaje_crudo.get("payload", {})
    
    if payload.get("canal") == "noticias_tech":
        titular = payload.get("titular", "Sin titular")
        print(f"[{nombre_nodo}] 📰 Nueva noticia interceptada: {titular}")

def main():
    print(f"Iniciando {nombre_nodo}...")
    cliente = NexusClient()
    cliente.connect(nombre_nodo)
    print(f"✅ {nombre_nodo} conectado. Escuchando el canal 'noticias_tech'...")
    
    cliente.listen(al_recibir_noticia)
    
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print(f"\nApagando {nombre_nodo}...")
        cliente.disconnect()

if __name__ == "__main__":
    main()

File: archivo-historico/ejemplos/nerve/02-pub-sub-basico/suscriptor.rs
use nerve::{NexusClient, Message};
use std::env;
use std::sync::mpsc;

fn main() {
    let args: Vec<String> = env::args().collect();
    let nombre_nodo = if args.len() > 1 {
        args[1].clone()
    } else {
        "suscriptor_default".to_string()
    }
    
    // Usar un Arc para compartir el nombre_nodo con el closure si es necesario, 
    // pero podemos capturarlo clonándolo
    let nodo_clone = nombre_nodo.clone();
    
    println!("Iniciando {}...", nombre_nodo);
    let mut cliente = NexusClient::new();
    cliente.connect(&nombre_nodo).expect("Error conectando");
    println!("✅ {} conectado. Escuchando el canal 'noticias_tech'...", nombre_nodo);

    let al_recibir_noticia = move |mensaje_crudo: Message| {
        if let Some(payload) = mensaje_crudo.payload {
            if let Some(canal) = payload.get("canal") {
                if canal == "noticias_tech" {
                    let titular = payload.get("titular").and_then(|t| t.as_str()).unwrap_or("Sin titular");
                    println!("[{}]  Nueva noticia interceptada: {}", nodo_clone, titular);
                }
            }
        }
    };

    cliente.listen(al_recibir_noticia);

    let (tx, rx) = mpsc::channel();
    ctrlc::set_handler(move || tx.send(()).expect("Error enviando señal"))
        .expect("Error al configurar ctrl+c");

    rx.recv().expect("Error esperando señal");
    
    println!("\nApagando {}...", nombre_nodo);
    cliente.disconnect();
}
